text/0000-regex-1.0.md

```diff
@@ -0,0 +1,984 @@
+
+  /// Returns an iterator of tuples, where each tuple is the name of the
+  /// capturing group (if it exists) and the matched text. If a capturing group
+  /// did not participate in a match, then the second element of the tuple is
+  /// None.
+  pub fn iter_named<'c>(&'c self) -> SubCapturesNamedIter<'c, 't>;
+
+  /// Returns the number of captured groups. This is always at least 1, since
+  /// the first unnamed capturing group corresponding to the entire match
+  /// always exists.
+  pub fn len(&self) -> usize;
+
+  /// Expands all instances of $name in the text given to the value of the
+  /// corresponding named capture group. The expanded string is written to
+  /// dst.
+  ///
+  /// The name in $name may be integer corresponding to the index of a capture
+  /// group or it can be the name of a capture group. If the name isn't a valid
+  /// capture group, then it is replaced with an empty string.
+  ///
+  /// The longest possible name is used. e.g., $1a looks up the capture group
+  /// named 1a and not the capture group at index 1. To exert more precise
+  /// control over the name, use braces, e.g., ${1}a.
+  ///
+  /// To write a literal $, use $$.
+  pub fn expand(&self, replacement: &str, dst: &mut String);
+}
+```
+
+The `Captures` type impls `Debug`, `Index<usize>` (for numbered capture groups)
+and `Index<str>` (for named capture groups). A downside of the `Index` impls is
+that the return value is bounded to the lifetime of `Captures` instead of the
+lifetime of the actual text searched because of how the `Index` trait is
+defined. Callers can work around that limitation if necessary by using an
+explicit method such as `at` or `name`.
+
+## Replacer
+[replacer]: #replacer
+
+The `Replacer` trait is a helper trait to make the various `replace` methods on
+`Regex` more ergonomic. In particular, it makes it possible to use either a
+standard string as a replacement, or a closure with more explicit access to a
+`Captures` value.
+
+```rust
+pub trait Replacer {
+  /// Appends text to dst to replace the current match.
+  ///
+  /// The current match is represents by caps, which is guaranteed to have a
+  /// match at capture group 0.
+  ///
+  /// For example, a no-op replacement would be
+  /// dst.extend(caps.at(0).unwrap()).
+  fn replace_append(&mut self, caps: &Captures, dst: &mut String);
+
+  /// Return a fixed unchanging replacement string.
+  ///
+  /// When doing replacements, if access to Captures is not needed, then
+  /// it can be beneficial from a performance perspective to avoid finding
+  /// sub-captures. In general, this is called once for every call to replacen.
+  fn no_expansion<'r>(&'r mut self) -> Option<Cow<'r, str>> {
+    None
+  }
+}
+```
+
+Along with this trait, there is also a helper type, `NoExpand` that implements
+`Replacer` like so:
+
+```rust
+pub struct NoExpand<'t>(pub &'t str);
+
+impl<'t> Replacer for NoExpand<'t> {
+    fn reg_replace(&mut self, _: &Captures) -> Cow<str> {
```

Comment:
Gah, I must have copied the wrong impl by accident. That is actually the current `0.1` trait. I'll fix it soon.